doc/sphinx/proof-engine/ltac.rst

```diff
@@ -713,339 +823,538 @@ tactic
+               assert (p := eq_refl 1 : X=1);    (* To filter the case X=1 *)
+               let x:= context f[O] in assert (x=O) (* To observe the context *)
+             end.
+           Goal True.
+           f (3+4).
+
+   .. todo There's a more realistic example from @JasonGross here:
+      https://github.com/coq/coq/pull/12103#discussion_r432996954
+
+   .. todo @term after context is actually @lconstr_pattern (term with holes), which is spliced away
+      Maybe worth restoring that distinction?
+
+   .. todo PR Didn't understand the following 2 paragraphs well enough to revise
+
+   Alternatively, when a metavariable in the form :n:`?id` occurs under binders,
    say :n:`x__1, …, x__n` and the expression matches, the
    metavariable is instantiated by a term which can then be used in any
    context which also binds the variables :n:`x__1, …, x__n` with
    same types. This provides with a primitive form of matching under
    context which does not require manipulating a functional term.
 
-   If the matching with :n:`@cpattern__1` succeeds, then :n:`@ltac_expr__1` is
-   evaluated into some value by substituting the pattern matching
-   instantiations to the metavariables. If :n:`@ltac_expr__1` evaluates to a
-   tactic and the match expression is in position to be applied to a goal
+   If the match expression is in position to be applied to a goal
    (e.g. it is not bound to a variable by a :n:`let in`), then this tactic is
    applied. If the tactic succeeds, the list of resulting subgoals is the
-   result of the match expression. If :n:`@ltac_expr__1` does not evaluate to a
-   tactic or if the match expression is not in position to be applied to a
-   goal, then the result of the evaluation of :n:`@ltac_expr__1` is the result
-   of the match expression.
-
-   If the matching with :n:`@cpattern__1` fails, or if it succeeds but the
-   evaluation of :n:`@ltac_expr__1` fails, or if the evaluation of
-   :n:`@ltac_expr__1` succeeds but returns a tactic in execution position whose
-   execution fails, then :n:`cpattern__2` is used and so on. The pattern
-   :n:`_` matches any term and shadows all remaining patterns if any. If all
-   clauses fail (in particular, there is no pattern :n:`_`) then a
-   no-matching-clause error is raised.
-
-   Failures in subsequent tactics do not cause backtracking to select new
-   branches or inside the right-hand side of the selected branch even if it
-   has backtracking points.
+   result of the match expression. If the match expression is not in position
+   to be applied to a goal, then the result of the evaluation of :n:`@ltac_expr__1`
+   is the result of the match expression.
+
+   .. exn:: Expression does not evaluate to a tactic.
+
+      :n:`@ltac_expr` must evaluate to a tactic.
 
    .. exn:: No matching clauses for match.
 
-      No pattern can be used and, in particular, there is no :n:`_` pattern.
+      None of the patterns match and, in particular, there is no :n:`_` pattern.
```

Comment:
Not technically true, consider
```coq
Goal True.
  match goal with
  | _ => fail
  | [ |- False ] => idtac
  end.
```

Reply:
The message appears when no branch matches the pattern whose action (initially) succeeds?

Reply:
That is closer, but consider
```coq
Goal True.
  multimatch goal with
  | _ => idtac
  | [ |- False ] => fail
  end; fail.
  (* Error: No matching clauses for match. *)
```
and
```coq
Goal True.
  lazymatch goal with
  | _ => fail
  | [ |- False ] => fail
  end.
  (* Error: Tactic failure. *)
```
I believe the precise/technical description may be:
The message appears when all successes of a `match` or `multimatch` have been exhausted before all potential matches/branches have been exhausted, and also occurs when no branch of a `lazymatch` has a matching pattern.

(This requires modelling each `match` as having multiple branches, each of which can have zero or more successes, and the evaluation of a match proceeds by going through each branch in turn, checking if it matches, and then exhausting the successes of that branch before moving on to the next one.)